### PR TITLE
proc: fix could not read process comm name error

### LINKED ISSUE #531: proc: Could not read process comm name when launching debugger
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)? b9b3e4ac03c1033fe6a8866813670fe58e96e0fd
2. What version of Go are you using? (`go version`)? 1.6.2
3. What operating system and processor architecture are you using?  linux (2.6.32-279.el6.x86_64)
4. What did you do? debugging
5. What did you expect to see? debugger started successfully
6. What did you see instead? 
Could not read process comm name: open /proc/29019/comm: no such file or directory

I think the cause of this issue is that /proc/[pid]/comm was not introduced before linux 2.6.33.

### PR DESCRIPTION
**Pull request by ggndnn**
fix #531

---
This change is [Reviewable](https://reviewable.io/reviews/derekparker/delve/532)

### CODE REVIEW

File: proc/proc_linux.go [hunk 1]
```diff
@@ -338,12 +339,32 @@ func (dbp *Process) loadProcessInformation(wg *sync.WaitGroup) {
 	defer wg.Done()
 
 	comm, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/comm", dbp.Pid))
-	if err != nil {
+	if(err == nil) {
+		// removes newline character
+		comm = comm[:len(comm)-1]
+	}
+	for (comm == nil || len(comm) <= 0) {
```
Comment (aarzilli) [inline proc/proc_linux.go]: This should be an if not a for, there is no way this loops more than once, right?

**Comment by aarzilli** [comment]
Commits need to be squashed, also there's a refactoring in progress, I don't know what @derekparker wants to do with this in the short term.

**Comment by ggndnn** [comment]
I get it, no problem, and commit has been squashed.

File: proc/proc_linux.go (hunk 1 above)
Comment (ggndnn) [inline proc/proc_linux.go]: Right, logically it's if. I meant to use break statement to leave current block when error occur.

**Comment by derekparker** [comment]
@ggndnn thanks for the fix! I'm going to hold off on merging this until https://github.com/derekparker/delve/pull/525 lands (should be soon). I'll make sure to ping you once that happens so we can get this merged.

**Comment by ggndnn** [comment]
@derekparker ok, no problem, thanks.

File: proc/proc_linux.go [hunk 2]
```diff
@@ -338,12 +339,32 @@ func (dbp *Process) loadProcessInformation(wg *sync.WaitGroup) {
 	defer wg.Done()
 
 	comm, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/comm", dbp.Pid))
-	if err != nil {
+	if err == nil {
+		// removes newline character
+		comm = comm[:len(comm)-1]
```
Comment (derekparker) [inline proc/proc_linux.go]: Please use https://golang.org/pkg/bytes/#TrimSuffix for this instead.

File: proc/proc_linux.go [hunk 3]
```diff
@@ -338,12 +339,32 @@ func (dbp *Process) loadProcessInformation(wg *sync.WaitGroup) {
...
+	}
+	for comm == nil || len(comm) <= 0 {
```
Comment (derekparker) [inline proc/proc_linux.go]: I don't understand the loop here, can we do without it?

**Comment by derekparker** [comment]
@ggndnn I'm going to break that large refactor into chunks so I'd like to merge your code after the feedback is taken care of. Thanks!

**Comment by ggndnn** [comment]
@derekparker I have commited my modifications. Thanks.

File: proc/proc_linux.go [hunk 4]
```diff
@@ -337,12 +339,30 @@ func (dbp *Process) loadProcessInformation(wg *sync.WaitGroup) {
...
+		comm = bytes.TrimSuffix(comm, []byte("\n"))
+	}
+
+	if comm == nil || len(comm) <= 0 {
+		stat, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/stat", dbp.Pid))
+		if err != nil {
+			fmt.Printf("Could not read proc stat: %v\n", err)
+			os.Exit(1)
+		}
+		expr := fmt.Sprintf("%d\\s*\\((.*)\\)", dbp.Pid)
```
Comment (derekparker) [inline proc/proc_linux.go]: This would be easier to read if we simply used `scan` here instead of using a regexp.

Comment (ggndnn) [inline proc/proc_linux.go]: I am not sure how to scan proc name including space character, any suggestion?

Comment (derekparker) [inline proc/proc_linux.go]: Sorry for delay on responding, vacation + conferences and such. However, you're right, I misspoke. Scanning is not the way to go here.

**Comment by derekparker** [comment]
@ggndnn if you can squash your commits I will merge. Thanks for the patch!

**Comment by ggndnn** [comment]
@derekparker commits have been squashed. Thanks!

File: proc/proc_linux.go [hunk 5]
```diff
@@ -350,12 +352,30 @@ func (dbp *Process) loadProcessInformation(wg *sync.WaitGroup) {
...
+		rexp, err := regexp.Compile(expr)
+		if err != nil {
+			fmt.Printf("Regexp compile error: %v\n", err)
+			os.Exit(1)
+		}
+		match := rexp.FindSubmatch(stat)
+		if match == nil {
+			fmt.Printf(fmt.Sprintf("No match found using regexp '%s' in /proc/%d/stat", expr, dbp.Pid))
```
Comment (derekparker) [inline proc/proc_linux.go]: Should of caught this before, but the `Printf(Sprintf())` is unnecessary, can just format with `Printf` directly. Also, needs newline at end of string.

Comment (ggndnn) [inline proc/proc_linux.go]: Yes, it's my fault. Thanks.

**Comment by ggndnn** [comment]
@derekparker modification has been committed.

**Pull request merged**